### PR TITLE
Implementation of minimax.

### PR DESCRIPTION
**Pull request by miselico**

### CODE REVIEW

**Comment by miselico** [comment]
I am getting more convinced that the implementation works. It is better than random, and plays on par with rdeep with a lot of samples. With a reduced number of samples for rdeep, this bot plays better.

**Comment by miselico** [comment]
Comparing the output of the minimax bot with rdeep (sampling rate 32 and rollout depth 5) in phase 2 resulted in the same choice in ~65% of the cases. This sounds low, but it might be that there are several equivalent choices.

**Comment by tae898** [comment]
```python
class RandMiniMaxBot(Bot):
    def __init__(self) -> None:
        self.bot_phase1 = RandBot(42)
        self.bot_phase2 = MiniMaxBot()

    def get_move(self, state: PlayerPerspective, leader_move: Optional[Move]) -> Move:
        if state.get_phase() == GamePhase.ONE:
            return self.bot_phase1.get_move(state, leader_move)

            # return self.delegate_phase2.get_move(state, leader_move)
        elif state.get_phase() == GamePhase.TWO:
            return self.bot_phase2.get_move(state, leader_move)
        else:
            raise ValueError("Phase ain't right.")


engine = SchnapsenGamePlayEngine()
bot1 = RandMiniMaxBot()
bot2 = RandBot(43)

winners = {"RandMiniMaxBot": 0, "RandBot": 0}
for i in tqdm(range(1000)):
    winner, _, _ = engine.play_game(bot1, bot2, random.Random(i))

    if winner == bot1:
        winners["RandMiniMaxBot"] += 1
    elif winner == bot2:
        winners["RandBot"] += 1
    else:
        raise ValueError

winners
```

```sh
{'RandMiniMaxBot': 668, 'RandBot': 332}
```

`RandMiniMaxBot` plays random in the phase 1, and plays minimax in the phase 2. It plays against the `RandBot`.

**Comment by tae898** [comment]
```python
class RdeepMiniMaxBot(Bot):
    def __init__(self) -> None:
        self.bot_phase1 = RdeepBot(
            num_samples=16, depth=4, rand=random.Random(42)
        )
        self.bot_phase2 = MiniMaxBot()

    def get_move(self, state: PlayerPerspective, leader_move: Optional[Move]) -> Move:
        if state.get_phase() == GamePhase.ONE:
            return self.bot_phase1.get_move(state, leader_move)
        elif state.get_phase() == GamePhase.TWO:
            return self.bot_phase2.get_move(state, leader_move)
        else:
            raise ValueError("Phase ain't right.")


engine = SchnapsenGamePlayEngine()
bot1 = RandMiniMaxBot()
bot2 = RdeepBot(num_samples=16, depth=4, rand=random.Random(43))

winners = {"RdeepMiniMaxBot": 0, "RdeepBot": 0}
for i in tqdm(range(1000)):
    winner, _, _ = engine.play_game(bot1, bot2, random.Random(i))

    if winner == bot1:
        winners["RdeepMiniMaxBot"] += 1
    elif winner == bot2:
        winners["RdeepBot"] += 1
    else:
        raise ValueError

winners
```

```sh
{'RdeepMiniMaxBot': 252, 'RdeepBot': 748}
```

`RdeepMiniMaxBot` plays rdeep in the phase 1, and plays minimax in the phase 2. It plays against the `RdeepBot`.

File: tests/test_bots.py [hunk 1]
```diff
@@ -24,3 +31,77 @@ def setUp(self) -> None:
...
+        self.bot_phase2 = MiniMaxBot()
+
+    def get_move(self, state: PlayerPerspective, leader_move: Optional[Move]) -> Move:
+        if state.get_phase() == GamePhase.ONE:
+            return self.bot_phase1.get_move(state, leader_move)
+        elif state.get_phase() == GamePhase.TWO:
+            return self.bot_phase2.get_move(state, leader_move)
+        else:
+            raise ValueError("Phase ain't right.")
+
```
Comment (miselico) [inline tests/test_bots.py]: @tae898 We would need these unit tsts to be much more fine grained. Basically, we should only have this Bot and call methods (get_move) on it directly. We then provide it with the needed arguments and checks whether it is giving us the move we expect.

**Comment by tae898** [comment]
@miselico

**Pull request merged**